testing/mock/middleware.go

```diff
@@ -0,0 +1,192 @@
+package mock
+
+import (
+	"bytes"
+	"strings"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	capabilitytypes "github.com/cosmos/ibc-go/modules/capability/types"
+	clienttypes "github.com/cosmos/ibc-go/v8/modules/core/02-client/types"
+	channeltypes "github.com/cosmos/ibc-go/v8/modules/core/04-channel/types"
+	porttypes "github.com/cosmos/ibc-go/v8/modules/core/05-port/types"
+	host "github.com/cosmos/ibc-go/v8/modules/core/24-host"
+	"github.com/cosmos/ibc-go/v8/modules/core/exported"
+)
+
+var _ porttypes.Middleware = (*BlockUpgradeMiddleware)(nil)
+
+// BlockUpgradeMiddleware does not implement the UpgradeableModule interface
+type BlockUpgradeMiddleware struct {
+	appModule *AppModule
+	IBCApp    *IBCApp // base application of an IBC middleware stack
+}
+
+// NewIBCModule creates a new IBCModule given the underlying mock IBC application and scopedKeeper.
+func NewBlockUpgradeMiddleware(appModule *AppModule, app *IBCApp) BlockUpgradeMiddleware {
+	appModule.ibcApps = append(appModule.ibcApps, app)
+	return BlockUpgradeMiddleware{
+		appModule: appModule,
+		IBCApp:    app,
+	}
+}
```

Comment:
Might be possible to embed the `IBCModule` app, passing it to the constructor instead of `app *IBCApp` and reduce code duplication that way. We should only need to implement the additional middleware methods. We could additional provide overrides for them in the `IBCApp` struct.

It may be beneficial to rename `IBCApp` to `IBCAppOverrides` or something similar, and reuse this struct for overridable middlewares as well.

Reply:
Looks like struct embedding actually defeats the purpose as then the struct would implement the app upgrade interface 😆

Reply:
no good deed goes unpunished! :D

Reply:
What if we create a new struct that both the `BlockUpgradeMiddleware` and the mock `IBCModule` are composed of that just contain the duplicated fns?

pinging @damiannolan since this is in a resolved thread!